### PR TITLE
feat: Implemented validate and sync rule function for sync-oscal-content command

### PR DESCRIPTION
**Pull request by RichardXuan**
## Description

Implemented validate and sync rule function for sync-oscal-content cac-control command

Sample cmd:

```shell
poetry run trestlebot sync-oscal-content cac-control --branch main --cac-content-root ~/content --committer-name axuan --committer-email axuan@redhat.com --dry-run --repo-path ~/trestlebot-workspace --product ocp4
```

Fixes CPLYTM-543, CPLYTM-540

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How has this been tested?

- [x] Local test
- [x] Unit test

**Test Configuration**:

- Firmware version:
- Hardware:
- Toolchain:
- SDK:

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: trestlebot/tasks/sync_oscal_content_task.py [hunk 1]
```diff
@@ -198,15 +274,13 @@ def _handle_controls_field(self, controls_data: List[Dict[str, Any]]) -> None:
...
+            oscal_control_id = self.catalog_helper.get_id(control["id"])
+
+            if oscal_control_id not in self.implemented_requirement_dict:
                 continue
 
-            # get rules field
-            rules = populate_if_dict_field_not_exist(control, "rules", [])
-            self._update_selections_rules_in_memory(rules, add_params=False)
+            oscal_control = self.implemented_requirement_dict[oscal_control_id]
+            self._update_control_file_change_in_memory(control, oscal_control)
```
[the remark this answers is not in the document]
Comment (RichardXuan) [reply trestlebot/tasks/sync_oscal_content_task.py:287]: From what i understand. It's expected. Every control file is a policy, if a policy change, the corresponding control need to change.

[the remark this answers is not in the document]
Comment (RichardXuan) [reply trestlebot/tasks/sync_oscal_content_task.py:287]: cc @marcusburghardt

**Comment by huiwangredhat** [comment]
Thanks for the PR. It needs updates.
I tested the feature with the prod rhel9 and profile anssi_bp28_minimal.profile.
1. generate the component-definition
2. remove the rules in component-definition
3. Check the anssi.yml, the rules haven't been removed

File: trestlebot/tasks/sync_oscal_content_task.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (marcusburghardt) [reply trestlebot/tasks/sync_oscal_content_task.py:287]: For product specific control files, such as `cis_rhel9`, `ccn_rhel9` or `nist_ocp4` it should be fine to update the control file because it would impact one single product. However, for product agnostic control files such as `anssi`, `pcidss` and `hipaa` the control file should be as generic as possible and each product would refine the rules and parameters on their profiles. Updating these product agnostic profiles based on a single product preference could create a situation impacting other products consuming it on their profiles. Unfortunately, I am not sure if we can safely determine if a control file is product specific or product agnostic at this moment before discussing the cases with CaC maintainers.

**Comment by RichardXuan** [comment]
> Thanks for the PR. It needs updates. I tested the feature with the prod rhel9 and profile anssi_bp28_minimal.profile.
> 
> 1. generate the component-definition
> 2. remove the rules in component-definition
> 3. Check the anssi.yml, the rules haven't been removed

@huiwangredhat Discussed result of this scenario https://docs.google.com/document/d/1WmWTQEnjSKKCuIxjz81uEte_vHwc3DcDcc5MB7lqaiU/edit?tab=t.0#heading=h.uvmlcz7h2s5v

File: trestlebot/tasks/sync_oscal_content_task.py [hunk 2]
```diff
@@ -148,16 +179,63 @@ def write_ordered_data_to_yaml(
...
+        In memory update cac control file changes
+        """
+        rule_list = populate_if_dict_field_not_exist(cac_control, "rules", [])
+
+        removed_variable = []
+        cac_rule_list = []
+        for rule_index, rule in enumerate(rule_list):
+            if "=" in rule:
+                # variable
+                v_id, v_value = rule.split("=")
```
Comment (marcusburghardt) [inline trestlebot/tasks/sync_oscal_content_task.py:195]: `v_value` is not used here. Maybe the value could be ignored: `v_id, _ = rule.split("=")`

File: trestlebot/tasks/sync_oscal_content_task.py [hunk 3]
```diff
@@ -148,16 +179,63 @@ def write_ordered_data_to_yaml(
...
+        # add rule
+        for rule in set(oscal_control_rules).difference(set(cac_rule_list)):
+            if rule in self.all_rule_ids_from_cac:
+                rule_list.append(rule)
+            else:
+                logger.warning(
+                    f"rule {rule} not exists in cac content repo: {self.cac_content_root}"
+                )
+
+    def _update_profile_change_in_memory(
```
Comment (marcusburghardt) [inline trestlebot/tasks/sync_oscal_content_task.py:226]: There are similar code between `_update_profile_change_in_memory` and `_update_control_file_change_in_memory`. Have you considered to refactor in orders to use smaller functions for common code shared between other functions? For example, to parse variables.

**Review by marcusburghardt — COMMENTED** [review]
I still would like to do some tests with this review tomorrow but can already share some comments to consider in the meantime. They are for now more about wording consistency in the documentation and some opportunities to simplify with small refactoring.

File: trestlebot/tasks/sync_oscal_content_task.py (hunk 2 above)
Comment (RichardXuan) [reply trestlebot/tasks/sync_oscal_content_task.py]: Updated

File: trestlebot/tasks/sync_oscal_content_task.py (hunk 3 above)
Comment (RichardXuan) [reply trestlebot/tasks/sync_oscal_content_task.py:239]: Already refactor this section. Make common code as function `_parse_single_variable`

**Comment by huiwangredhat** [comment]
For the refactored code, LGTM.
I tested two scenarios:
**Sync removed the rule from oscal to cac profile and control file**
1.Generate the component-definition for ocp4 with high-rev-4.profile
2.Delete the rules in component-definition: version_detect_in_ocp and routes_protected_by_tls
3.Run sync-oscal-cac
Result: 
The rule version_detect_in_ocp should be deleted in the profile high-rev-4.profile
The rule routes_protected_by_tls which is associated with the valid control-ids should be deleted in the control file nist_ocp4

**Sync added rule from oscal to cac profile and control file**
1.Restored the component-definition from the last scenario 
2.Run sync-oscal-cac
3.Result: 
The rule version_detect_in_ocp hasn’t be included in  the high-rev-4.profile (This is the expected result, about the reason I asked from the [doc](https://docs.google.com/document/d/1WmWTQEnjSKKCuIxjz81uEte_vHwc3DcDcc5MB7lqaiU/edit?tab=t.0#heading=h.m4bjsm7w0rbl))
The rule routes_protected_by_tls  was added to control file nist_ocp4 for the associated control-id

**Review by huiwangredhat — COMMENTED** [review]
For the refactored code, LGTM.
I tested two scenarios:
**Sync removed the rule from oscal to cac profile and control file**
1.Generate the component-definition for ocp4 with high-rev-4.profile
2.Delete the rules in component-definition: version_detect_in_ocp and routes_protected_by_tls
3.Run sync-oscal-cac
Result: 
The rule version_detect_in_ocp should be deleted in the profile high-rev-4.profile
The rule routes_protected_by_tls which is associated with the valid control-ids should be deleted in the control file nist_ocp4

**Sync added rule from oscal to cac profile and control file**
1.Restored the component-definition from the last scenario 
2.Run sync-oscal-cac
3.Result: 
The rule version_detect_in_ocp hasn’t be included in  the high-rev-4.profile (This is the expected result, about the reason I asked from the [doc](https://docs.google.com/document/d/1WmWTQEnjSKKCuIxjz81uEte_vHwc3DcDcc5MB7lqaiU/edit?tab=t.0#heading=h.m4bjsm7w0rbl))
The rule routes_protected_by_tls  was added to control file nist_ocp4 for the associated control-id

**Review by huiwangredhat — APPROVED** [review]

**Review by marcusburghardt — APPROVED** [review]
This is relatively new and unfortunately we don't have enough real data to test. When we have, some corner cases may appear and we can expect that some code will need to be updated. We are actively working on this transformation now and I don't see any reason to block this PR. Let's merge it so we can include more features on top of it. Thanks for the work @RichardXuan
